CONTRIBUTING.md

```diff
@@ -0,0 +1,113 @@
+## How to Contribute
+
+The OCS-Operator project is under the [Apache 2.0 license](LICENSE). We accept
+contributions via GitHub pull requests. This document outlines how to
+contribute to the project.
+
+## Contribution Flow
+
+Developers must follow these steps to make a change:
+
+1. Fork the `openshift/ocs-operator` repository on GitHub.
+2. Create a branch from the master branch, or from a versioned branch (such
+   as `release-4.2`) if you are proposing a backport.
+3. Make changes.
+4. Create tests as needed and ensure that all tests pass.
+5. Ensure your commit messages include sign-off.
+6. Push your changes to a branch in your fork of the repository.
+7. Submit a pull request to the `openshift/ocs-operator` repository.
+8. Work with the community to make any necessary changes through the code
+   review process (effectively repeating steps 3-8 as needed).
+
+## Developer Environment Installation
+
+Instructions to create a dev environment for OCS-operator can be found in the
+[main project documentation](./README.md#Installation of development builds).
+
+## Commit Requirements
+
+In addition to passing [functional tests](./README.md#functional-tests), all
+commits to OCS-operator must follow the guidelines in this section. You can
+verify your changes meet most of these standards by running the following
+command in the repository's root directory:
+
+```
+make ocs-operator-ci
+```
+
+### Commits Per Pull Request
+
+*This requirement cannot be tested by make ocs-operator-ci.*
+
+OCS-operator is a project which maintains several versioned branches
+independently. When backports are necessary, monolithic commits make it
+difficult for maintainers to cleanly backport only the necessary changes.
+
+Where possible, pull requests should be composed of multiple commits that
+each make small but meaningful changes. Striking a balance between minimal
+commits and logically complete changes is an art as much as a science, but
+when it is possible and reasonable, divide your pull request into more commits.
```

Comment:
I suggested to have this discussion separately, but here are some thoughts:

A pull request (PR) presents a larger logical change that should be added in one.
A pull request should usually do one such larger change and only that, and it should do it completely.
If you want to change independent things, raise multiple PRs.

Commits in a PR are steps to achieve this larger change of the PR. 
Each commit should do one smaller step toward the larger goal, it should do only this step, but do it completely.
So usually a commit should compile and pass tests.
In particular, if a code change breaks unit or functional tests, these tests need to be adapted in the same commit.
Additional tests to test the new feature or api or so should be added in a separate commit in the same PR, after the code change.
Exceptions from the requirements of passing tests are that for reasons of easier backporting, we somtimes want to split, e.g. changes in the sources and results of the crd or csv generation into separate commits.

As mentioned above, what is deemed appropriate granularity is often to some extent a matter of personal style and taste, but there are some guidlines:

* unrelated formatting changes outside the context of the actual change should go into separate commits (and maybe separate PRs)
* preparing changes like refactorings or code rearrangements should be split into separate commits if possible. this can be done even after creating the eventual change by interactive rebasing.
* adding new functions and using them can (and often should) be done in separate commits (usually committing the new function along with their unit test)

...

Reply:
I want to have this discussion now. One of the main friction points with new(er) developers are our rigorous commit guidelines, so I would want these things written down in a first draft.

That said, almost :100: on everything you said. One contention:

> adding new functions and using them can (and often should) be done in separate commits (usually committing the new function along with their unit test)

I don't like commits that add unused code. Unit tests for that code don't count as usage by the compiled binary. I think new functions should accompany any code changes to call those functions as well.

Reply:
> I want to have this discussion now. One of the main friction points with new(er) developers are our rigorous commit guidelines, so I would want these things written down in a first draft.

Right. This is dearly needed!

> That said, almost 💯 on everything you said. 

Great! 😄 

> One contention:
> 
> > adding new functions and using them can (and often should) be done in separate commits (usually committing the new function along with their unit test)
> 
> I don't like commits that add unused code. Unit tests for that code don't count as usage by the compiled binary. I think new functions should accompany any code changes to call those functions as well.

I knew this could be highly controversial. 😄 
It's ok to skip it, it's mostly a taste thing of mine.

But let's hear others: @egafford , @umangachapagain , @raghavendra-talur , @nbalacha , ...

Reply:
I'm with @jarrpa on this one. The rest of what you said above makes a ton of sense, but backporting commits that introduce unused functions could very easily leave us with a fair bit of confusing, orphaned code.

Reply:
I like commits which introduce only the new function and then subsequent commits that call the function. This helps in backporting if the caller function goes through many code changes before the backport. Orphaned functions seem harmless to me. That said, it is only a preference.

Reply:
Thanks @raghavendra-talur - that's a good argument.
A new function should be added at least with a unit test covering it.

Reply:
I'm with @jarrpa.
I've recently seen some instances of having a commit with just the new function. I had to jump through the commits to find a caller so that I could make better sense of the function. Turns out it wasn't there and I was left more confused than ever.
It also makes reviewing difficult and could end up with a flawed function being committed.

Reply:
For me, it makes reviewing easier instead of more difficult if it is in separate commits. But that's just my poor brain that likes to have chunks as small as possible at a time to digest. 😄 

And yeah, usually the function should be used in the *next* commit of the patchset rather than in some arbitrary later commit. But the problem of a useless or unused function can be hidden in any commit even if we usually use the function in the same commit in which we introduce it. 

But as mentioned, I'm probably not going to fight very hard on this... ;-)